iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,386 @@
+    if size < MAX_HANDSHAKE_LENGTH {
+        // Skipping garbage
+        let mut buf = vec![0_u8; size];
+        let _ = stream.read_exact(&mut buf).await?;
+    } else {
+        return Err(new_error("Error reading client hello!"));
+    }
+    // And then we have clients public key
+    let mut key = [0_u8; 32];
+    let _ = stream.read_exact(&mut key).await?;
+    Ok(PublicKey(Vec::from(key)))
+}
+
+pub async fn send_client_hello(stream: &mut OwnedWriteHalf, key: &[u8]) -> io::Result<()> {
+    let garbage = generate_garbage();
+
+    #[allow(clippy::cast_possible_truncation)]
+    stream.write_u8(garbage.len() as u8).await?;
+    stream.write_all(&garbage).await?;
+    stream.write_all(key).await?;
+
+    Ok(())
+}
+
+pub async fn read_server_hello(stream: &mut OwnedReadHalf) -> Result<PublicKey, Error> {
+    let size = stream.read_u8().await? as usize;
+    if size < MAX_HANDSHAKE_LENGTH {
+        // Reading garbage
+        let mut buf = vec![0_u8; size];
+        let _ = stream.read_exact(&mut buf).await?;
+    } else {
+        return Err(new_error("Error reading server hello!"));
+    }
+    // Then we have clients public key
+    let mut key = [0_u8; 32];
+    let _ = stream.read_exact(&mut key).await?;
+    Ok(PublicKey(Vec::from(key)))
+}
+
+pub async fn send_server_hello(stream: &mut OwnedWriteHalf, key: &[u8]) -> io::Result<()> {
+    let garbage = generate_garbage();
+
+    #[allow(clippy::cast_possible_truncation)]
+    stream.write_u8(garbage.len() as u8).await?;
+    stream.write_all(&garbage).await?;
+    stream.write_all(key).await?;
+
+    Ok(())
+}
+
+pub async fn read_message(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u32().await? as usize;
+    if size > 0 && size < MAX_MESSAGE_LENGTH {
+        let mut buf = vec![0_u8; size];
+        let mut read = 0;
+        while read < size {
+            read += stream.read_exact(&mut buf[read..]).await?;
+        }
+
+        return Ok(Message(buf));
+    }
+    Err(new_error("Error reading message!"))
+}
+
+pub async fn send_message(stream: &mut OwnedWriteHalf, data: &[u8]) -> Result<(), Error> {
+    if data.len() > MAX_MESSAGE_LENGTH {
+        warn!(
+            "Message length exceeds maximum length of {}!",
+            MAX_MESSAGE_LENGTH
+        );
+        return Err(new_error("Message size is too big!"));
+    }
+    #[allow(clippy::cast_possible_truncation)]
+    let size: u32 = data.len() as u32;
+    let mut buf: Vec<u8> = Vec::with_capacity(data.len() + 2);
+    buf.write_u32(size).await?;
+    buf.write_all(data).await?;
+    stream.write_all(buf.as_slice()).await?;
+    stream.flush().await?;
+    Ok(())
+}
+
+fn generate_garbage() -> Vec<u8> {
+    let rng = &mut rand::thread_rng();
+    let mut garbage = vec![0_u8; rng.gen_range(64, 255)];
+    rng.fill_bytes(&mut garbage);
+    garbage
+}
+
+fn new_error(text: &str) -> io::Error {
+    io::Error::new(ErrorKind::Other, text)
+}
```

Comment:
This is kinda bad practice. This is clearly not an io error, rather this error inside of a networking layer. You should create some error type